File: org.tec.datastructures/src/org/tec/datastructures/hierarchical/AVLTree.java
package org.tec.datastructures.hierarchical;

import org.tec.datastructures.nodes.AVLTreeNode;

public class AVLTree<T extends Comparable<T>>{
	private AVLTreeNode<T> root;
	
	public AVLTree() {
		this(null);
	}
	
	public AVLTree(AVLTreeNode<T> Root) {
		setRoot(Root);
	}

	public AVLTreeNode<T> getRoot() {
		return root;
	}

	public void setRoot(AVLTreeNode<T> root) {
		this.root = root;
	}
	
	private Integer validateHeight(AVLTreeNode<T> node){
		if (node == null) {
			return 0;
		}
		return node.getHeight(); 
	}
	
	public void append(T value) {
		root = appendRecursive(value,root);
		
	}
	
	private AVLTreeNode<T> appendRecursive(T value, AVLTreeNode<T> node) {
		if (node == null) {
			node = new AVLTreeNode<T>(value);
		}
		else if(value.compareTo(node.getValue()) < 0) {
			node.setLeft(appendRecursive(value, node.getLeft()));
		}
		else if (value.compareTo(node.getValue()) > 0) {
			node.setRight(appendRecursive(value, node.getRight()));
		}
		node = balance(node);
		node.calculateHeight();
		return node;
	}
	
	public AVLTreeNode<T> search(T value){
		return searchRecursive(value,root);
	}
	
	private AVLTreeNode<T> searchRecursive(T value, AVLTreeNode<T> node) {
		if (node == null) {
			return null;
		}
		else if (node.getValue().compareTo(value) > 0) {
			return searchRecursive(value, node.getLeft());
		}
		else if (node.getValue().compareTo(value) < 0) {
			return searchRecursive(value, node.getRight());
		}
		return node;	
	}

	public void delete(T value) {
		deleteRecursive(value, root);
	}

	private AVLTreeNode<T> deleteRecursive(T value, AVLTreeNode<T> node) {
		
		if (node == null) {
			return null;
		}
		else if (node.getValue().compareTo(value) > 0) {
			node.setLeft(deleteRecursive(value, node.getLeft()));
		}
		else if (node.getValue().compareTo(value) < 0) {
			node.setRight(deleteRecursive(value, node.getRight()));
		} 
		else if (node.getLeft() != null && node.getRight() != null) {
			T replace = findMin(node.getRight()).getValue();
			
			node.setValue(replace);
			node.setRight(deleteRecursive(replace,node.getRight()));
		}
		else {
			if (node == root) {
				node = (node.getLeft() == null) ? node.getRight() : node.getLeft();
				root = node;
			}else {
				node = (node.getLeft() == null) ? node.getRight() : node.getLeft();
			}
			
			
		}
		
		node = balance(node);
		return node;
	}
	
	private AVLTreeNode<T> findMin(AVLTreeNode<T> node) {
		if (node.getLeft() == null) {
			return node;
		}
		return findMin(node.getLeft());
	}
	
	private AVLTreeNode<T> balance(AVLTreeNode<T> node) {
		
		if (node == null) {
			return null;
		}
		
		Integer balance = validateHeight(node.getLeft())
				- validateHeight(node.getRight());
		if (balance == 2){
			if (node.getLeft().getLeft() != null) {
				if (node == root){
					root = node.getLeft();
				}
				node = node.rotateLeftChild();
				
			}
			else {
				if (node == root){
					root = node.getLeft().getRight();
				}
				node = node.rotateDoubleLeft();
			}
		}
		else if (balance == -2) {
			if (node.getRight().getRight() != null) {
				if (node == root) {
					root = node.getRight();
				} 
				node = node.rotateRightChild();
			}
			else {
				if (node == root) {
					root = node.getRight().getLeft();
				}
				node = node.rotateDoubleRight();
			}
		}
		node.calculateHeight();
		return node;
	}
	
}

File: C#/org.tec.datastructures/AVLTree.cs
using System;
namespace org.tec.datastructures
{
    using System;

    public class AVLTree<T> where T : IComparable<T>
    {
        private AVLNode<T> root;

        public void Append(T value)
        {
            this.root = this.AppendAux(this.root, value);
        }

        private AVLNode<T> AppendAux(AVLNode<T> node, T value)
        {
            if (node == null)
            {
                node = new AVLNode<T>(value);
            }else if (node.value.CompareTo(value) == 1)
            {
                node.left = this.AppendAux(node.left, value);
            }else if (node.value.CompareTo(value) == -1)
            {
                node.right = this.AppendAux(node.right, value);
            }
            node = Balance(node);
            return node;
        }

        private AVLNode<T> Balance(AVLNode<T> node)
        {
            if (node == null)
            {
                return null;
            }
            node.calculateHeight();
            if (node.balance == 2)
            {
                if(node.left != null && node.left.balance == -1){
                    node = node.RotateLeftRight();
                }else{
                    node = node.RotateRight();
                }
            }
            else if (node.balance == -2)
            {
                if (node.left != null && node.left.balance == 1)
                {
                    node = node.RotateRightLeft();
                }
                else
                {
                    node = node.RotateLeft();
                }
            }
            node.calculateHeight();
            return node;
        }

        public void Delete(T value)
        {
            this.root = this.DeleteAux(this.root, value);
        }

        private AVLNode<T> DeleteAux(AVLNode<T> node, T value)
        {
            if (node != null){
                
            }else if (node.value.CompareTo(value) == 1)
                {
                    node.left = this.DeleteAux(node.left, value);
                }else if (node.value.CompareTo(value) == -1)
                {
                    node.right = this.DeleteAux(node.right, value);
                }else if ((node.left != null) && (node.right != null))
                {
                    node.value = this.FindMin(node.right).value;
                    node.right = this.DeleteAux(node.right, value);
                }
                node = (node.left == null) ? node.right : node.left;

            node = Balance(node);
            return node;
        }

        private AVLNode<T> FindMin(AVLNode<T> node)
        {
            if (node.left == null)
            {
                return node;
            }
            return this.FindMin(node.left);
        }

        public AVLNode<T> Get(T value) =>
            this.GetAux(this.root, value);

        private AVLNode<T> GetAux(AVLNode<T> node, T value)
        {
            if (node != null)
            {
                if (node.value.CompareTo(value) == -1)
                {
                    node = this.GetAux(node.left, value);
                    return node;
                }
                if (node.value.CompareTo(value) == 1)
                {
                    node = this.GetAux(node.right, value);
                }
            }
            return node;
        }

        public AVLNode<T> getRoot() =>
            this.root;

        private int Height(AVLNode<T> node)
        {
            if (node == null)
            {
                return 0;
            }
            return node.height;
        }
    }

    public class AVLNode<T>
    {
        public int balance;
        public int height;
        public AVLNode<T> left;
        public AVLNode<T> right;
        public T value;

        public AVLNode(T value)
        {
            this.value = value;
        }

        public void calculateHeight()
        {
            if (!this.hasChilden())
            {
                this.height = 0;
                this.balance = 0;
            }
            else
            {
                if (this.left == null)
                {
                    this.height = this.right.height + 1;
                    this.balance = -this.right.height;
                }
                else if (this.right == null)
                {
                    this.height = this.left.height + 1;
                    this.balance = this.left.height;
                }
                else
                {
                    this.height = Math.Max(this.left.height, this.right.height);
                    this.balance = this.left.height - this.right.height;
                }
            }
        }

        public bool hasChilden()
        {
            return left != null || right != null;
        }

        public AVLNode<T> RotateLeft()
        {
            AVLNode<T> right = this.right;
            this.right = right.left;
            right.left = (AVLNode<T>)this;
            this.calculateHeight();
            right.calculateHeight();
            return right;
        }

        public AVLNode<T> RotateLeftRight()
        {
            this.left = this.left.RotateLeft();
            return this.RotateRight();
        }

        public AVLNode<T> RotateRight()
        {
            AVLNode<T> left = this.left;
            this.left = left.right;
            left.right = (AVLNode<T>)this;
            this.calculateHeight();
            left.calculateHeight();
            return left;
        }

        public AVLNode<T> RotateRightLeft()
        {
            this.right = this.right.RotateRight();
            return this.RotateLeft();
        }
    }
}
